### PR TITLE
relay: fix GTID recover after relay log recovered

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

MySQL may not save all previous GTID sets in its binlog file as `Previous_gtids` event in some special scenarios (not confirmed yet, but maybe a bug, ref <https://bugs.mysql.com/bug.php?id=93809>).

example:
- the real previous GTID sets: `00c04543-f584-11e9-a765-0242ac120002:1-60`
- GTID sets in `Previous_gtids` event: `00c04543-f584-11e9-a765-0242ac120002:45-60`
- then the recovered GTID sets will become `00c04543-f584-11e9-a765-0242ac120002:45-60`, but this should still be `00c04543-f584-11e9-a765-0242ac120002:1-60`.
- use `00c04543-f584-11e9-a765-0242ac120002:45-60` to request the master is wrong, because this will cause the master to send binlog events in `00c04543-f584-11e9-a765-0242ac120002:1-44`. if this events are purged, an error (`ERROR 1236 (HY000): The slave is connecting using CHANGE MASTER TO MASTER_AUTO_POSITION = 1, but the master has purged binary logs containing GTIDs that the slave requires.`) will reported; if this events are not purged, they will be sent again.

### What is changed and how it works?

only truncate the `Stop` of the GTID set intervals, like truncating `00c04543-f584-11e9-a765-0242ac120002:1-60` with `00c04543-f584-11e9-a765-0242ac120002:40-50` will get `00c04543-f584-11e9-a765-0242ac120002:1-50`.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@amyangfei @WangXiangUSTC PTAL

File: pkg/gtid/gtid.go [hunk 1]
```diff
@@ -173,6 +179,36 @@ func (g *mySQLGTIDSet) Contain(other Set) bool {
...
+func (g *mySQLGTIDSet) Truncate(end Set) error {
+	if end == nil {
+		return nil // do nothing
+	}
+	if !g.Contain(end) {
+		return terror.ErrGTIDTruncateInvalid.Generate(g, end)
+	}
+	endGs := end.(*mySQLGTIDSet) // already verify the type is `*mySQLGTIDSet` in `Contain`.
+	if endGs == nil {
+		return nil // do nothing
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/gtid/gtid.go]: `end` is an `interface`, and the instance it referencing can still be `nil`.

like 
```golang
var end *mySQLGTIDSet
Truncate(end)
```

File: pkg/gtid/gtid_test.go [hunk 2]
```diff
@@ -178,3 +180,185 @@ func (s *testGTIDSuite) TestMairaGTIDContain(c *C) {
...
+			after:  "00c04543-f584-11e9-a765-0242ac120002:100",
+		},
+		{
+			before: "00c04543-f584-11e9-a765-0242ac120002:40-60",
+			end:    "00c04543-f584-11e9-a765-0242ac120002:45-55",
+			after:  "00c04543-f584-11e9-a765-0242ac120002:40-55",
+		},
+		{
+			before: "00c04543-f584-11e9-a765-0242ac120002:40-60:70:80-100",
+			end:    "00c04543-f584-11e9-a765-0242ac120002:45-55:70:85-95",
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/gtid/gtid_test.go]: 👍 added in https://github.com/pingcap/dm/pull/335/commits/feeed29c79316e9149bf5cb43ad473bab76190dc.

[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/gtid/gtid_test.go]: and how about also add for mariadb

[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/gtid/gtid_test.go]: how about also add for mariadb

File: pkg/gtid/gtid.go [hunk 3]
```diff
@@ -173,6 +179,36 @@ func (g *mySQLGTIDSet) Contain(other Set) bool {
...
+	}
+
+	for sid, setG := range g.set.Sets {
+		setE, ok := endGs.set.Sets[sid]
+		if !ok {
+			continue // no need to truncate for this SID
+		}
+		for i, interG := range setG.Intervals {
+			for _, interE := range setE.Intervals {
+				if interG.Start <= interE.Start && interG.Stop >= interE.Stop {
```
Comment (amyangfei) [inline pkg/gtid/gtid.go]: Does the MariaDB has the same recover mechanism with MySQL, in order to be safe to re-sync from the new GTID

File: relay/relay.go [hunk 4]
```diff
@@ -387,6 +390,11 @@ func (r *Relay) handleEvents(ctx context.Context, reader2 reader.Reader, transfo
 					cfg := r.cfg.From
 					r.tctx.L().Error("the requested binlog files have purged in the master server or the master server have switched, currently DM do no support to handle this error",
 						zap.String("db host", cfg.Host), zap.Int("db port", cfg.Port), log.ShortError(err))
+					// log the status for debug
+					pos, gs, err2 := utils.GetMasterStatus(r.db, r.cfg.Flavor)
+					if err2 == nil {
+						r.tctx.L().Info("current master status", zap.Stringer("position", pos), zap.Stringer("GTID sets", gs))
```
Comment (amyangfei) [inline relay/relay.go]:
```suggestion
						r.tctx.L().Info("current master status", zap.Stringer("position", pos), log.WrapStringerField("GTID sets", gs))
```

File: relay/relay.go [hunk 5]
```diff
@@ -344,6 +344,9 @@ func (r *Relay) tryRecoverLatestFile(parser2 *parser.Parser) error {
 		if result.Recovered {
 			r.tctx.L().Warn("relay log file recovered",
 				zap.Stringer("from position", latestPos), zap.Stringer("to position", result.LatestPos), log.WrapStringerField("from GTID set", latestGTID), log.WrapStringerField("to GTID set", result.LatestGTIDs))
+			if err = latestGTID.Truncate(result.LatestGTIDs); err != nil {
+				return err
+			}
 			err = r.meta.Save(result.LatestPos, result.LatestGTIDs)
```
Comment (amyangfei) [inline relay/relay.go]: should save meta with `LatestGTID`?

File: relay/relay.go (hunk 4 above)
Comment (csuzhangxc) [reply relay/relay.go]: done in https://github.com/pingcap/dm/pull/335/commits/9611a0c1c71645fe6b77aaaadb1ea19c1136e7b7.

File: relay/relay.go (hunk 5 above)
Comment (csuzhangxc) [reply relay/relay.go]: you are right! thanks a lot. done in https://github.com/pingcap/dm/pull/335/commits/9611a0c1c71645fe6b77aaaadb1ea19c1136e7b7.

**Comment by lance6716** [comment]
The problem in PR description is caused by MySQL slave forget to set `log_slave_updates=ON`. The GTIDs when `log_slave_updates` is OFF will not be saved in `Previous_gtids` events.

```
MySQL [(none)]> select @@global.gtid_purged;
+------------------------------------------+
| @@global.gtid_purged                     |
+------------------------------------------+
| 250e86b4-ba95-11ea-9e12-b15aac539905:1-3 |
+------------------------------------------+
1 row in set (0.000 sec)

MySQL [(none)]> select @@global.gtid_executed;
+------------------------------------------+
| @@global.gtid_executed                   |
+------------------------------------------+
| 250e86b4-ba95-11ea-9e12-b15aac539905:1-4 |
+------------------------------------------+

MySQL [(none)]> show binlog events in 'binlog.000003';
+---------------+-----+----------------+-----------+-------------+-------------------------------------------------------------------+
| Log_name      | Pos | Event_type     | Server_id | End_log_pos | Info                                                              |
+---------------+-----+----------------+-----------+-------------+-------------------------------------------------------------------+
| binlog.000003 |   4 | Format_desc    |         1 |         123 | Server ver: 5.7.23-log, Binlog ver: 4                             |
| binlog.000003 | 123 | Previous_gtids |         1 |         154 |                                                                   |
| binlog.000003 | 154 | Gtid           |         2 |         219 | SET @@SESSION.GTID_NEXT= '250e86b4-ba95-11ea-9e12-b15aac539905:4' |
| binlog.000003 | 219 | Query          |         2 |         282 | BEGIN                                                             |
| binlog.000003 | 282 | Table_map      |         2 |         327 | table_id: 108 (gset.t1)                                           |
| binlog.000003 | 327 | Write_rows     |         2 |         367 | table_id: 108 flags: STMT_END_F                                   |
| binlog.000003 | 367 | Xid            |         2 |         398 | COMMIT /* xid=6 */                                                |
+---------------+-----+----------------+-----------+-------------+-------------------------------------------------------------------+
7 rows in set (0.001 sec)
```

since now we introduced #1430 , the `Truncate` interface is no longer needed.

**Pull request merged**